numpy/lib/arraysetops.py

```diff
@@ -332,14 +338,45 @@ def intersect1d(ar1, ar2, assume_unique=False):
     >>> from functools import reduce
     >>> reduce(np.intersect1d, ([1, 3, 4, 3], [3, 1, 2, 1], [6, 3, 4, 2]))
     array([3])
+    
+    To return the indices of the values common to the input arrays
+    along with the intersected values:
+    
+    >>> np.intersect1d([1,2,3,4], [2,1,4,6], return_indices=True)
+    (array(1,2,4), array([0,1,3]), array([1,0,2]))
     """
-    if not assume_unique:
+    if not (assume_unique or return_indices):
         # Might be faster than unique( intersect1d( ar1, ar2 ) )?
+        # Added the extra conditional because np.unique uses an additional 
+        # keyword when return_indices is specified for intersect1d. 
         ar1 = unique(ar1)
         ar2 = unique(ar2)
-    aux = np.concatenate((ar1, ar2))
+
+
+    if return_indices:
+      if not assume_unique:
+        a1, ind1 = unique(ar1, return_index = True) #a1 not necessary
+        a2, ind2 = unique(ar2, return_index = True)
+        aux = np.concatenate([a1,a2])
+      else:
+        aux = np.concatenate([ar1, ar2])
```

Comment:
Can these two branches be made the same? Ie, just re-use the name `ar1` to refer to the unique'd `ar1`, something like:

```
# make arrays unique if needed
if not assume_unique:
    if return_indices:
        ar1, ind1 = unique(ar1, return_index=True)
        ar2, ind2 = unique(ar2, return_index=True)
    else:
        ar1 = unique(ar1)
        ar2 = unique(ar2)
aux = np.concatenate([ar1, ar2])

# now special case
if return_indices:
   # the code you have below
```